### PR TITLE
Track BackupEngine and BackupStorageHandle errors together.

### PR DESCRIPTION
**Pull request by ajm188**
This addresses #6349.

This change requires `BackupHandle` to implement the `concurrency.ErrorRecorder` interface in order to guarantee that errors that happen either in the backup engine code _or_ in the backup storage handle code end up in the same place. This provides a single place for builtinbackupengine to check for errors before proceeding with writing the manifest.

This also required adding `ErrorRecorder` to certain BackupStorage implementations that didn't previously need them (`filebackupstorage`, `gcsbackupstorage`), and allowed me to turn simplify some calls (i.e. `bh.errors.HasErrors()` becomes `bh.HasErrors()`) in other implemtations (`s3backupstorage`, `azblobbackupstorage`, `cephbackupstorage`).

Finally, I added a test for the builtinbackupengine+s3backupstorage error case, which required me to change the typing of `S3BackupHandle.client` from `*s3.S3` to the recommended `s3iface.S3API` which is just the interface that `*s3.S3` implements.

### CODE REVIEW

File: go/vt/mysqlctl/builtinbackupengine.go [hunk 1]
```diff
@@ -287,19 +286,28 @@ func (be *BuiltinBackupEngine) backupFiles(ctx context.Context, params BackupPar
...
-	if rec.HasErrors() {
-		return rec.Error()
+
+	// BackupHandles now support the ErrorRecorder interface for tracking errors
+	// across any goroutines that fan out to take the backup. This means that we
+	// can discard the local error recorder and put everything through the bh.
+	//
+	// This mitigates the scenario where bh.AddFile() encounters an error asynchronously,
+	// which ordinarily would be lost in the context of `be.backupFile`, i.e. if an
+	// error were encoutered
```
Comment (deepthi) [inline go/vt/mysqlctl/builtinbackupengine.go:307]:
```suggestion
	// BackupHandle supports the ErrorRecorder interface for tracking errors
	// across any goroutines that fan out to take the backup. This means that we
	// don't need a local error recorder and can put everything through the bh.
	//
	// This handles the scenario where bh.AddFile() encounters an error asynchronously,
	// which ordinarily would be lost in the context of `be.backupFile`, i.e. if an
	// error were encountered
```

File: go/vt/mysqlctl/s3backupstorage/s3_test.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	}
+
+	return &req, &s3.PutObjectOutput{}
+}
+
+func TestAddFileError(t *testing.T) {
+	bh := &S3BackupHandle{client: &s3ErrorClient{}, readOnly: false}
+	wc, err := bh.AddFile(aws.BackgroundContext(), "somefile", 100000)
+
+	if err != nil {
```
Comment (deepthi) [inline go/vt/mysqlctl/s3backupstorage/s3_test.go:29]: can you change all the `t.Errorf` calls to use `require.NoError`?

File: go/vt/mysqlctl/s3backupstorage/s3_test.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	}
+
+	if err := wc.Close(); err != nil {
+		t.Errorf("TestAddFile() could not close writer, got %s", err)
+		return
+	}
+
+	bh.waitGroup.Wait() // wait for the goroutine to finish, at which point it should have recorded an error
+
+	if !bh.HasErrors() {
```
Comment (deepthi) [inline go/vt/mysqlctl/s3backupstorage/s3_test.go:52]: use `assert.True`.

File: go/vt/mysqlctl/s3backupstorage/s3_test.go (hunk 2 above)
Comment (ajm188) [reply go/vt/mysqlctl/s3backupstorage/s3_test.go]: of course!

File: go/vt/mysqlctl/s3backupstorage/s3_test.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+	return &req, &s3.PutObjectOutput{}
+}
+
+func TestAddFileError(t *testing.T) {
+	bh := &S3BackupHandle{client: &s3ErrorClient{}, readOnly: false}
+
+	wc, err := bh.AddFile(aws.BackgroundContext(), "somefile", 100000)
+	require.NoErrorf(t, err, "AddFile() expected no error, got %s", err)
+	assert.NotEqual(t, nil, wc, "AddFile() expected non-nil WriteCloser")
```
Comment (deepthi) [inline go/vt/mysqlctl/s3backupstorage/s3_test.go:32]: There's also an assert.NotNil, but I'll accept this.

**Pull request merged**